Improve stepper driver PWM support

dot-bob opened the pull request:
I have added a menu option to the lcd display which allows editing of the stepper PWM current.
I have added the ability to save the PWM current to EEPROM.

Both the Ultimaker 2 and Wanaho Duplicator 6 use PWM to set the stepper driver current.

Commit:
Added the ability to edit stepper digipot current from the display.
Support saving the stepper digipot current to eeprom.

Commit:
Added the ability to edit stepper PWM current from the display.
Support saving the stepper PWM current to eeprom.
Rebase and clean up pull request.

dot-bob committed:
Update configuration_store.cpp

dot-bob committed:
Revert some changes

Remove incomplete digipot changes and focus only on PWM

Commit:
Merge branch 'digipot_save' of https://github.com/dot-bob/Marlin into digipot_save

Bob-the-Kuhn commented:
Will you be expanding the menu system to include all the systems that can set the motor current via firmware?

dot-bob commented:
I based the menu changes off what was already there with DAC_STEPPER_CURRENT.  DAC stepper current uses a percentage where the PWM is set in mA.  A standard menu system for setting stepper current is a good idea but may be out of scope of this pull request.  Right now I am trying to add basic LCD menu and save support right now.  
I based most of the functionality off existing functions and strings to limit the scope of the pull request.

I would be open to the discussion of a standard menu for stepper current, perhaps under a different pull request or new issue.

thinkyhead commented on Marlin/configuration_store.cpp:
```diff
@@ -626,6 +630,15 @@ void MarlinSettings::postprocess() {
+      const unsigned long dummy_motor_current_setting[3] = {0, 0, 0};
+      for (uint8_t q = 3; q--;)
+        EEPROM_WRITE(dummy_motor_current_setting[q]);
+    #endif
```
You can just do this here:
```cpp
dummy = 0.0;
for (uint8_t q = 3; q--;)
  EEPROM_WRITE(dummy);
```

thinkyhead commented on Marlin/stepper.cpp:
```diff
@@ -1470,6 +1474,9 @@ void Stepper::report_positions() {
       const uint8_t digipot_ch[] = DIGIPOT_CHANNELS;
       digitalPotWrite(digipot_ch[driver], current);
     #elif HAS_MOTOR_CURRENT_PWM
+      if((driver >= 0)&& (driver<3))
```
Formatting quibble:
```cpp
if (WITHIN(driver, 0, 2))
```

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -1249,6 +1254,23 @@ void kill_screen(const char* lcd_msg) {
 
+  #if HAS_MOTOR_CURRENT_PWM
+    void update_motor_power() {
+	  stepper.digipot_current(0, stepper.motor_current_setting[0]);
```
Watch out for tabs. One sneaked in here.

thinkyhead commented:
Overall a strong first showing — bravissimo! A few cosmetic changes are all that's needed. Once you've made those tweaks, I can merge everything as a single neat commit.

Commit:
Misc cleanups and removed stray tab.
Changed pins_MINIRAMBO.h and pins_ULTIMAIN_2.h to use a default MOTOR_CURRENT_PWM_RANGE if it isn't defined else where.

Commit:
Misc cleanups and removed stray tab.
Changed pins_MINIRAMBO.h and pins_ULTIMAIN_2.h to use a default MOTOR_CURRENT_PWM_RANGE if it isn't defined else where.

dot-bob commented:
I made the changes above and I also changed the files pins_MINIRAMBO.h and pins_ULTIMAIN_2.h to only use a default MOTOR_CURRENT_PWM_RANGE if it isn't defined else where.  This will allow someone to define it in the configuration.h or configuration_adv.h and use it to scale the PWM_MOTOR_CURRENT for their specific hardware.  So if they define a current of 1300 will actually equal 1300 mA to the hardware.

Commit:
Fix compile error

Commit:
Update to latest bugfix-1.1.x as of 06-08-17

Commit:
Merge remote-tracking branch 'origin/bugfix-1.1.x' into digipot_save

Commit:
Fix build error due to upstream changes.
Modify the Travis CL script to test PWM_MOTOR_CURRENT.

Andrei-Pozolotin commented:
+1

thinkyhead commented:
Squashed and cleaned up at #7136.
